adding a few extension development items

choldgraf opened the pull request:
This PR adds some short sections to the developer documentation. It addresses a few items in #7044 .

There are 2 main things that this PR does:

* Show how to develop extensions that depend on other extensions
* Show how to define functions that can be used in Jinja

closes https://github.com/sphinx-doc/sphinx/issues/7044

choldgraf commented:
It's not totally clear to me, but I don't *think* that the travis errors are related to this, but please let me know if they are and I can fix!

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
 
 
+Extension How-tos
+-----------------
```
It would be better move such tips to sub page. A top page of extension developing is not good for them. How about moving them to individual file like `extdev/tips.rst` or `extdev/howto.rst`?

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+incorrect configuration in the template.
+
+To define your own template function, you'll need to define two functions inside
+your module, a **parent function** and a **template function**.
```
These functions are not parent and child. So it would be better to name **parent function** to **page context event handler** or **registration function**.

choldgraf replied:
OK - I updated the names, let me know how it looks

tk0miya replied:
LGTM!

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+To define your own template function, you'll need to define two functions inside
+your module, a **parent function** and a **template function**.
+
+First, define the **parent function**, which accepts the
```
It would be better to emphasis the keyword only once to me. How about highlighting on first appearance or declaration?

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+Within the parent function, define
+the **template function** that you'd like to use within Jinja. The template function
+should return a string or docutils object that Sphinx uses in the templating process
+(it will have access to all of the variables that are passed to the parent function).
```
Could you fold a text by (almost) 80 chars? Our checker warns this line overs it a bit.
https://travis-ci.org/sphinx-doc/sphinx/jobs/641448161?utm_medium=notification&utm_source=github_status

Note: It checks by 85 chars. But 80 chars is better.

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+
+Within the parent function, define
+the **template function** that you'd like to use within Jinja. The template function
+should return a string or docutils object that Sphinx uses in the templating process
```
In Jinja context, nobody wants docutils object, I guess. Is there useful case?

choldgraf replied:
oops good catch, I meant a Python object, I'll clarify in the next push

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+First, define the **parent function**, which accepts the
+arguments for :event:`html-page-context`.
+
+Within the parent function, define
```
You don't have to define template functions inside event handler. I feel defining them on module level is simple and easy for test.

choldgraf replied:
Interesting - I assumed that this was necessary in order to have the template function get access to all of the variables that are passed to the event handler function (e.g., `app`, `context`, etc). Is that not the case?

choldgraf replied:
I just confirmed that (in another project), when I defined the function outside of the event handler function, then the template function no longer had access to the `context` variable

tk0miya replied:
Sure. You need to define it inside handler function if you need to access it. Okay, let's go as is.

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+.. code:: python
+
+   def setup(app):
+      app.setup_extension("recommonmark")
```
Personally I prefer to 4spaces indentation. Please add one more space here :-)

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+      # This is a configuration that you've specified for users in `conf.py`
+      js_variable = app.config['my_javascript_variable']
+      js_text = "var my_variable = '%s';" % js_variable
+      app.add_js_file(None, body=js_text)
```
This block is also be indented by 4spaces.

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+
+   # We connect this function to the step after the builder is initialized
+   def setup(app):
+      app.connect('builder-inited', add_js_variable)
```
How about adding `app.add_config_value('my_javascript_variable')`? It makes this example copy-pastable!

tk0miya commented on doc/extdev/index.rst:
```diff
@@ -189,6 +189,123 @@ as metadata of the extension.  Metadata keys currently recognized are:
+site's :file:`conf.py` file, but this is not available to you as an extension
+developer.
+
+.. module:: sphinx.application
```
Could you add `:noindex:` option to here please?

tk0miya reviewed (changes requested):
LGTM with nits!

choldgraf commented:
Thanks for the feedback!

I've addressed most of the comments - with one exception about function nesting for template functions. I also did two extra things:

1. Re-worded the beginning paragraph of the "extensions" page to make it a bit easier to follow (this included moving a section into the new "howto" page because it seemed to make more sense in there than at the top of the page)
2. Added a section to the file-wide metadata about how to use these fields within a template, closing out the last item in https://github.com/sphinx-doc/sphinx/issues/7044

Let me know what you think!

jorisvandenbossche commented on doc/extdev/index.rst:
```diff
@@ -3,54 +3,47 @@
+linking resources that Sphinx uses in the build process (like CSS or HTML
+files), and notifying Sphinx of everything the extension offers (such
+as directive or role definitions). The ``app`` argument is an instance of
+:class:`.Sphinx` and gives you control over most aspects of the Sphinx build.
```
Is the `.` in ``` :class:`.Sphinx` ``` correct?

choldgraf replied:
I _think_ so? I noticed that it was used elsewhere in the docs so I just copied it over here, e.g.: https://github.com/sphinx-doc/sphinx/blob/e628afd5cd0c565f1e33abb01cac26180455182f/sphinx/transforms/__init__.py#L55

tk0miya replied:
Yes, it's correct.

jorisvandenbossche reviewed (commented):
Really nice additions, thanks @choldgraf !

tk0miya commented on doc/extdev/howto.rst:
```diff
@@ -0,0 +1,161 @@
+Finally, in your extension's ``setup()`` function, add your registration
+function as a callback for :event:`html-page-context`.
+
+.. code:: python
```
Could you use `code-block` instead?

tk0miya commented on doc/extdev/howto.rst:
```diff
@@ -0,0 +1,161 @@
+First, define the registration function, which accepts the
+arguments for :event:`html-page-context`.
+
+Within the registration function, define
```
Could you fold the line by (almost) 80 chars? it seems this line folded strangely.

tk0miya replied:
Abve line is also.

tk0miya commented on doc/extdev/howto.rst:
```diff
@@ -0,0 +1,161 @@
+
+Sometimes it is useful to define your own function in Python that you wish to
+then use in a template. For example, if you'd like to insert a template value
+with logic that depends on the user's configuration in the site, or if you'd
```
In Sphinx, we call whole of documents as "project" or "document". "site" is also good. But it is not appropriate on building PDF. I prefer to call it "project".

tk0miya commented on doc/usage/restructuredtext/basics.rst:
```diff
@@ -258,8 +258,8 @@ target formats (HTML, LaTeX) have a limited supported nesting depth.
 
-Field lists (:duref:`ref <field-lists>`) are sequences of fields marked up like
-this::
+Field lists (:duref:`docutils ref <field-lists>`) are sequences of fields
```
We usually use "(ref)" link in this page. It would be better to change them all in once. Anyway, it should be in another PR.

choldgraf replied:
If you tell me what you'd like to see I am happy to do it here, to be honest I don't understand the difference between `duref` and `ref` so I was just trying to copy what else was there

tk0miya commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,60 @@ fields marked up like this::
+
+.. note::
+
+    The *values* of field lists will be parsed as
```
Where this emphasis comes from?

tk0miya commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,60 @@ fields marked up like this::
     :fieldname: Field content
 
-Sphinx provides custom behavior for bibliographic fields compared to docutils.
+Field lists are :duref:`originally defined in docutils <field-lists>`
```
This section describes about "Bibliographic fields" in Sphinx. So it should refer this section on reST spec. 
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#bibliographic-fields

choldgraf replied:
is there a way for me to do this with rST? I'm not sure what is the right syntax...

stephenfin replied:
I think you want

    :duref:`ref <bibliographic-fields>`

tk0miya commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,60 @@ fields marked up like this::
+.. note::
+
+    A field list appearing after the document title *will* be part of the
+    *docinfo* as normal and will be displayed in the output.
```
This is incorrect. *docinfo* (a.k.a. bibliographic fields) is placed at top of the document except invisible elements (ex. comments). Surely "appearing after the document title" is correct. That's not all.

In addition, *docinfo* is a special naming for field-lists at the top. So it should be called as mere field-list if placed middle of document.

choldgraf replied:
Thanks for your thoughts - I am a little confused and wonder if you could help clarify. The reason I assumed that "field lists at the top are not for the docinfo in Sphinx" is because of this text [from this page](https://www.sphinx-doc.org/en/master/usage/restructuredtext/field-lists.html#file-wide-metadata):

> A field list near the top of a file is normally parsed by docutils as the docinfo which is generally used to record the author, date of publication and other metadata. However, in Sphinx, a field list preceding any other markup is moved from the docinfo to the Sphinx environment...a field list appearing after the document title will be part of the docinfo as normal and will be displayed in the output.

I believe the "note" text isn't an addition by me, it is already in the docs, I just made it a note.

choldgraf replied:
I am still not sure how to resolve this one

tk0miya replied:
Unfortunately, the current documentation is wrong.

The spec of reST format describesthe bibliographic fields (a.k.a docinfo) as following:

>When a field list is the first non-comment element in a document (after the document title, if there is one), it may have its fields transformed to document bibliographic data. This bibliographic data corresponds to the front matter of a book, such as the title page and copyright page.
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#bibliographic-fields

This means a special field-list placed at the top of the document becomes a bibliographic fields (a.k.a. docinfo). And other field-lists are mere field-lists. So "a field list appearing after the document title" is not a docinfo.

tk0miya commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,60 @@ fields marked up like this::
+Any key / value pairs in field lists that are placed *before* the page's
+title will be available to the Jinja template when building the page within
+the :data:`meta` attribute. For example, if a page had the following text
+before its first title:
```
It would be better to separate mark-ups and how-to write themes. This might confuse the readers.

choldgraf replied:
Where do you think would be a better place for this information? And is there any other case that a user would use field lists other than for their use within a theme?

stephenfin replied:
I think this would be best placed in the HTML themes development document.

tk0miya reviewed (commented):
Thank you for update! I reviewed again. Could you update this again please?
It might be better to separate this PR to multiple PRs if you'd like to go step by step. (Either is okay to me)
Thanks,

choldgraf commented:
Hey - I can try another iteration or two and if it's too much then we can split PRs, but I'm happy to give a shot at merging the whole thing first :-)

I had a couple quick questions for you on your comments!

stephenfin commented on doc/extdev/howto.rst:
```diff
@@ -0,0 +1,170 @@
+Extension How-tos
```
I'd like to see this document split up. I think there are at least two separate documents here: one related to development of themes for the HTML builder (for the custom template functions and injection of Javascript sections), a builders document (for the entry point discussion), and a general overview document (for the inter-extension dependency documentation).

I'd also like to see these documents go into the `doc/development` directory as that's where everything in the `extdev` directory should be moved eventually.

choldgraf replied:
ok I'll give this a shot at splitting into 3 files

stephenfin commented on doc/extdev/index.rst:
```diff
@@ -3,54 +3,47 @@
+activates the extension by placing the extension's module name
+(or a sub-module) in their :confval:`extensions` configuration value.
+
+When a Sphinx site is built, Sphinx will attempt to import each
```
I wouldn't talk about Sphinx sites, since there are multiple builders, not all of them HTML. Instead, how about

  When :program:`sphinx-build` is executed, ...

stephenfin commented on doc/theming.rst:
```diff
@@ -20,6 +20,10 @@ the theme's look and feel.
 Themes are meant to be project-unaware, so they can be used for different
 projects without change.
 
+.. note::
```
This entire document should be moved to the development directory, and the information on injecting Javascript files included there.

choldgraf replied:
Do you want me to do that as a part of this PR?

choldgraf replied:
I'll move it and see what you think

stephenfin commented on doc/usage/restructuredtext/basics.rst:
```diff
@@ -258,8 +258,8 @@ target formats (HTML, LaTeX) have a limited supported nesting depth.
-Field lists (:duref:`ref <field-lists>`) are sequences of fields marked up like
-this::
+Field lists (:ref:`docutils ref <field-lists>`) are sequences of fields
+marked up like this::
```
Could this be done as a separate PR? I don't think there's a dependency here, right?

stephenfin commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,60 @@ fields marked up like this::
+.. note::
+
+    The values of field lists will be parsed as
+    strings, you cannot use Python collections such as lists or dictionaries.
```
Replace comma with full stop

stephenfin reviewed (changes requested):
This is really good. Some comments inside, mostly related to the long-term plans to move everything in `doc/extdev` to separate documents in `doc/development`. Let me know if you've any questions.

choldgraf commented:
Thanks for the feedback @stephenfin - I've tried to address each of your comments in the latest push. It looks like we're out-of-sync with master, but I'll hold off on a rebase until others have had a chance to give feedback.

I can try to put some cycles into getting this merged quickly if you've got the time to iterate on it!

stephenfin commented:
> I can try to put some cycles into getting this merged quickly if you've got the time to iterate on it!

Sounds good to me

choldgraf commented:
How do you recommend proceeding on this one? It seems like some of the merge conflicts are for content that has been moved in this PR? It's a bit unclear to me though.

stephenfin commented:
> How do you recommend proceeding on this one? It seems like some of the merge conflicts are for content that has been moved in this PR? It's a bit unclear to me though.

I'd suggest rebasing and force pushing. It'll mess up history a little (GitHubPRs--) but it's probably the cleanest way to do this. Also, you should propose against the `3.x` branch since `master` is for the next major version and this is perfectly okay to include in the next minor release. You'll have to change this once you rebase and push by selecting the dropdown at the top of this page under the title.

choldgraf commented:
ah ok - alright I'll give it a shot today

stephenfin commented:
> ah ok - alright I'll give it a shot today

Great. Let me know if you run into any issues.

choldgraf committed:
several documentation improvements to theming, templating, and extension development

choldgraf committed:
flake8

choldgraf commented:
OK, I've just committed these changes on top of `3.x` and changed the target branch. I think I've also addressed all of the comments, and got the rearranging of content where we want it to be. Please let me know if I've missed something in the rebase etc!

choldgraf commented:
Hey all - anything else for me to take care of here? Maybe ping @stephenfin ?

choldgraf commented:
Hi all - just a note that in 1-4 weeks time I will be starting paternity leave, and will no longer be able to put in significant time to this PR. I've rebased and moved branches, and I believe I've addressed the comments above. It would be great to get this in before I drop off the earth for a bit.

choldgraf commented on doc/development/tutorials/builders.rst:
```diff
@@ -0,0 +1,32 @@
+
+.. versionadded:: 1.6
+
+:term:`Builder` extensions can be discovered by means of `entry points`_ so
```
```suggestion
:term:`builder` extensions can be discovered by means of `entry points`_ so
```

choldgraf committed:
Update doc/development/tutorials/builders.rst

stephenfin commented on doc/development/tutorials/theming-dev.rst:
```diff
@@ -1,7 +1,5 @@
-.. highlight:: python
```
As above, this isn't really a tutorial. Could we place it in `doc/development` rather than `doc/development/tutorials`? Also, drop the `-dev` suffix - it's obvious based on the folder it's in

stephenfin commented on doc/development/tutorials/theming-dev.rst:
```diff
@@ -154,6 +160,137 @@ templating to put the color options into the stylesheet.  When a documentation
+            return "Your string is %s" % mystring
+        # Add it to the page's context
+        context['my_func'] = my_func
+    # Your extension's setup function
```
nit: can you add a newline between these two functions?

stephenfin commented on doc/development/tutorials/theming-dev.rst:
```diff
@@ -154,6 +160,137 @@ templating to put the color options into the stylesheet.  When a documentation
+   </div>
+
+
+Inject javsacript based on user configuration
```
JavaScript

stephenfin commented on doc/development/tutorials/theming-dev.rst:
```diff
@@ -154,6 +160,137 @@ templating to put the color options into the stylesheet.  When a documentation
+First, you may append ``_t`` to the end of any static files included with your
+extension. This will cause Sphinx to process these files with the templating
+engine, allowing you to embed variables and control behavior. See
+:ref:`theming-static-templates` for more information.
```
An example `ls`-style directory listing here would be super useful. You could use the [ASCII box-drawing characters](https://en.wikipedia.org/wiki/Box-drawing_character) for really nice looking output.

choldgraf replied:
I've added this in the latest commit, lemme know what you think

stephenfin commented on doc/development/tutorials/builders.rst:
```diff
@@ -0,0 +1,32 @@
+Configuring builders
```
Can you move this up a level? It's not really a tutorial, though it is development stuff

stephenfin commented on doc/development/tutorials/builders.rst:
```diff
@@ -0,0 +1,32 @@
+dotted name of the extension module. Here is an example of how an entry point
+for 'mybuilder' can be defined in the extension's ``setup.py``::
+
+    setup(
```
I know this is mostly copy-pasted, but can we add

    .. code-block:: python

?

stephenfin commented on doc/development/tutorials/overview.rst:
```diff
@@ -0,0 +1,32 @@
+activate another extension at run-time, ensuring that you have access to its
+functionality.
+
+For example, the following code activates the "recommonmark" extension:
```
Can we use

    ``recommonmark``

instead?

stephenfin reviewed (changes requested):
This is almost there. Couple of file moves and some typo fixes and I think we can merge this. Thanks for the hard work here!

stephenfin commented:
> Hi all - just a note that in 1-4 weeks time I will be starting paternity leave, and will no longer be able to put in significant time to this PR. I've rebased and moved branches, and I believe I've addressed the comments above. It would be great to get this in before I drop off the earth for a bit.

Apologies for the delay getting back to this. Contributing to Sphinx is a hobby that I don't get to spend as much time on as I'd like :disappointed: As noted above though, this is almost there and it should be possible close it out this week.

PS: Congrats, and good luck! :smile:

choldgraf committed:
rearranging a few docs and small updates

choldgraf commented:
Thanks for the comments! I think that I've addressed them in the latest commit, lemme know what you think!

(and thanks for the well-wishes!)

tk0miya commented on doc/development/theming.rst:
```diff
@@ -154,6 +160,157 @@ templating to put the color options into the stylesheet.  When a documentation
+There are two ways to inject variables into the JavaScript space based on user
+configuration.
+
+First, you may append ``_t`` to the end of any static files included with your
+extension. This will cause Sphinx to process these files with the templating
+engine, allowing you to embed variables and control behavior.
```
This behavior is not promised in public. And I'm still hesitating to support it officially (because it's not a "static" file!). So it should not be documented.

stephenfin replied:
Ah, so this is an implementation detail? I think we should either document this or remove this capability. If we do neither, [Hyrum's Law](https://www.hyrumslaw.com/) comes into play.

tk0miya replied:
Personally, I'd like to remove it ASAP. But it has behaved so since its beginning (maybe). And some extensions implicitly depend on the behavior. So we need to consider how to remove it safely.

choldgraf replied:
I believe it is documented in public here: https://www.sphinx-doc.org/en/master/theming.html#static-templates

that's where I found out about this feature, this PR is just trying to make it more discoverable. I guess I could remove this addition, but I feel like then it should be _removed_ from the documentation if it is explicitly not a desired feature.

tk0miya replied:
Okay, it's my wrong. I was thinking about `html_static_path`. But this is another topic. Please forget my comment above. As you quoted, files in the static directory of a theme are treated as a Jinja template and rendered on copying. That's a public feature.

What I described is `html_static_path`. It is also treated files under the specified directory as a Jinja template, but it is not official feature.

tk0miya commented on doc/development/theming.rst:
```diff
@@ -154,6 +160,157 @@ templating to put the color options into the stylesheet.  When a documentation
+extension. This will cause Sphinx to process these files with the templating
+engine, allowing you to embed variables and control behavior.
+
+For example, the following JavaScript structure:
+
+.. code-block:: bash
+
+   mymodule/
+   ├── _static
+   │   └── myjsfile.js_t
+   └── mymodule.py
+
+Will result in the following static file placed in your HTML's build output:
```
AFAIK, some additional code is needed to copy the `myjsfile.js_t`. As a well-known hack, many extensions appends their own static directory to the `html_static_path` on the `config-inited` event (or similar one).
```
def append_html_static_path(app, config):
    config.html_static_path.append('path/to/my/_static')

def setup(app):
    app.connect('config-inited', append_html_staitc_path)
```

But I feel modifying `html_static_path` is a bit hacky. So I prefer to copy them in their own code explicitly:
```
from sphinx.util.fileutil import copy_asset_file

def copy_custom_files(app, exc):
    if app.builder.format == 'html' and not exc:
        staticdir = path.join(app.builder.outdir, '_static')
        copy_asset_file('path/to/my/_static/myjsfile.js_t', staticdir)

def setup(app):
    app.connect('builder-inited', copy_custom_files)
```

stephenfin replied:
Yes, this is good information. Perhaps we could include this too?

choldgraf replied:
I'll add a short section that describes this

tk0miya replied:
Please forget my comment here also. The files under static directory under a theme will be copied automatically. No additional code is needed.

tk0miya commented on doc/usage/restructuredtext/field-lists.rst:
```diff
@@ -9,19 +9,38 @@ fields marked up like this::
     :fieldname: Field content
 
-Sphinx provides custom behavior for bibliographic fields compared to docutils.
+Field lists are :duref:`originally defined in docutils <bibliographic-fields>`
```
Incorrect. Field-lists do not have a special responsibility to show information about the page. They are only one of document structures like lists, tables, images, and so on.

I think this sentence should mention to the bibliographic fields instead. The bibliographic fields (a.k.a. docinfo) is a special field list that is placed at the top of the document. And they have a special responsibility to describe the page.

stephenfin replied:
Okay, yeah, this is confusing now. Good spot. Previously, we talked about field lists in general and then went on to talk about the special "file-wide metadata" behaviour that kicks in when they are at the top of the file. Now we're muddling them up. Could we keep this section before "File-wide metadata" mostly as-is, just adding the note about field lists being parsed as strings? The rework you've done of the next section makes sense, IMO

tk0miya replied:
+1 for keeping this section as-is.

choldgraf replied:
OK - will remove the first sentence and fix this

stephenfin commented:
Thanks for the review, @tk0miya :+1:

choldgraf committed:
comments

choldgraf commented:
OK I believe that I've addressed the latest comments in the last commit. Let me know if there's anything else I should tackle (e.g. if you want me to just remove all of the `_t` stuff)

tk0miya reviewed (approved):
LGTM!

stephenfin commented on doc/development/tutorials/overview.rst:
```diff
@@ -0,0 +1,32 @@
+Developing extensions overview
```
This document is still in the wrong place, IMO. Could we move it up a level? As with the builders document, this is general reference material, not tutorial-style content

stephenfin replied:
Just for reference, my long running doc rework is based on how [Django do their documentation](https://jacobian.org/2009/nov/10/what-to-write/) (that's a blog from one of the founders of the project, iirc). This document falls into the topical guide category, IMO

choldgraf replied:
just did this in the latest commit 👍

stephenfin reviewed (changes requested):
One last change and I'll be happy, I promise 🙈

choldgraf committed:
moving extension dev overview

choldgraf commented:
OK, I think all comments addressed!

stephenfin commented:
Thanks!

tk0miya commented:
Congrats!
